### PR TITLE
[FIX][17.0] account: fix error when delete the SO on form view

### PR DESCRIPTION
**Pull request by SonCrits**
[[BUG][17.0] viin_sales_unlink- Xảy ra lỗi khi xóa SO](https://viindoo.com/web#id=51865&cids=1&menu_id=289&model=viin.helpdesk.ticket&view_type=form)
-

This PR
-

- Fix bug when delete SO

**Step**
- Create SO
- Confirm SO, add delivered If the product is a warehouse product
- Create Invoice
- Delete Invoice
- Cancel SO
- Delete SO

**Output**

- Displays the error totals.subtotals order is not iterable

**Reason**

- When unlink SO, tax value no longer exists, so, return value of totals is false, and when try loop totals.subtotals_order (undefine value) will be display this error

**Solution**

- Check the total input value. If there is no value or the return value is false, return it

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by casp-odoo — CHANGES_REQUESTED** [review]
I can't reproduce the issue on the latest runbot 17.0. Maybe some more specific steps are not mentioned ?

**Comment by SonCrits** [comment]
https://github.com/odoo/odoo/assets/95740163/3de2082f-061f-4663-a0f4-e0806c5f20d7

@casp-odoo , I did it again on the latest runbot, giving you an attached video as proof

**Review by casp-odoo — COMMENTED** [review]
OK, my bad : I tested with Enterprise. The bug only occurs in Community though, which might give further clue about what's actually happening. Might be worth investigating a bit ! 

I unfortunately have no further ideas on the how-to ; let's see if @Levizar is more enlightened.

**Comment by SonCrits** [comment]
@Levizar Do you have any solution for this problem?

**Comment by SonCrits** [comment]
> OK, my bad : I tested with Enterprise. The bug only occurs in Community though, which might give further clue about what's actually happening. Might be worth investigating a bit !
> 
> I unfortunately have no further ideas on the how-to ; let's see if @Levizar is more enlightened.

thank you @casp-odoo

**Comment by SonCrits** [comment]
@Levizar It seems you are missing notifications in this PR. Ping

**Comment by Levizar** [comment]
Hi @SonCrits, no I didn't miss it.

I know it's a bit long to wait, sorry for that. 
I'm a bit drawn under PRs to review, was sick last week and today is an official holiday in Belgium.

I'll have a look to this one tomorrow.
Have a nice day

File: addons/account/static/src/components/tax_totals/tax_totals.js [hunk 1]
```diff
@@ -150,6 +150,9 @@ export class TaxTotalsComponent extends Component {
 
     formatData(props) {
         let totals = JSON.parse(JSON.stringify(toRaw(props.record.data[this.props.name])));
+        if (!totals) {
+			return;
+		}
```
Comment (duong77476-viindoo) [inline addons/account/static/src/components/tax_totals/tax_totals.js:155]:
```suggestion
        if (!totals) {
            return;
        }
```

Comment (duong77476-viindoo) [reply addons/account/static/src/components/tax_totals/tax_totals.js:155]: Need to preserve the format i suppose

**Review by Levizar — CHANGES_REQUESTED** [review]
First of all, thank you for your PR.
I investigated a lot on this today and needed to ask for several colleagues input/help.

Regarding the PR itself:
1) As @duong77476-viindoo said, could you format it?  
2) could you update your commit message to include the minimal step? (it is shorter than  currently described) and to include the following explanation?

Reproduction steps:
- disable any debugger parameter that would on in the url
- create a new sale order
- add a sale order line with a tax
- save (and don't leave the record!)
- delete the sale order

Regarding the bug, there are several issues happening at the same time:
- Business code issue:
    - Our code wasn't defensive and we were assuming the computed value would always be set to the computed value.
- Framework JS issue: (not OWL itself but its integration into Odoo)
    - When the record is deleted, a re-rendering gets triggered.
    - As we deleted the record in the back, the browsed record in the front-end gets its data reset to the default values a field could have. Because the default value of a binary field is `False`, `tax_totals` gets set to `false` for the reset record.

Now, we have a bunch of race conditions going on.
In the invoicing view, it doesn't happen because the rendering is slower and it gets cancelled before being done. Thus the field `tax_totals` hasn't the chance to re-render itself and thus, to call `TaxTotalsComponent.formatData` through the call of `onWillRender`.
In the case of the sale order view, the field has the time to gets rendered, the method gets called and the traceback happens.

Activating the debug parameter is slowing the rendering enough for the rendering to get canceled.

In short terms (stable and master), the business code should be more defensive and be able to handle the fields default values.
In the long term, the framework JS team will try to prevent the useless rendering to happens and may try to find a solution to prevent accidental rendering to crash.

**Comment by SonCrits** [comment]
@Levizar Thank you, I have updated the commit and formatted the code. Please help me review again

File: addons/account/static/src/components/tax_totals/tax_totals.js [hunk 2]
```diff
@@ -150,6 +150,9 @@ export class TaxTotalsComponent extends Component {
 
     formatData(props) {
         let totals = JSON.parse(JSON.stringify(toRaw(props.record.data[this.props.name])));
+        if (!totals) {
+        	return;
+        }
```
Comment (Levizar) [inline addons/account/static/src/components/tax_totals/tax_totals.js:155]:
```suggestion
        if (!totals) {
            return;
        }
```

**Review by Levizar — CHANGES_REQUESTED** [review]
Sorry to be this nitpicky but it is to keep consistency.

May I suggest you to setup a linter or a formatter?
This way, you could let it take care of it automatically.

**Review by Levizar — APPROVED** [review]
@robodoo r+
